src/viam/sdk/components/base/base.hpp

```diff
@@ -0,0 +1,86 @@
+/// @file components/base/base.hpp
+///
+/// @brief Defines a `Base` component.
+#pragma once
+
+#include <string>
+
+#include <viam/api/component/base/v1/base.pb.h>
+
+#include <viam/sdk/common/linear_algebra.hpp>
+#include <viam/sdk/common/proto_type.hpp>
+#include <viam/sdk/common/utils.hpp>
+#include <viam/sdk/config/resource.hpp>
+#include <viam/sdk/registry/registry.hpp>
+#include <viam/sdk/subtype/subtype.hpp>
+
+namespace viam {
+namespace sdk {
+
+/// @defgroup Base Classes related to the `Base` component.
+
+/// @class BaseSubtype
+/// @brief Defines a `ResourceSubtype` for the `Base` component.
+/// @ingroup Base
+class BaseSubtype : public ResourceSubtype {
+   public:
+    std::shared_ptr<ResourceServerBase> create_resource_server(
+        std::shared_ptr<SubtypeService> svc) override;
+    std::shared_ptr<ResourceBase> create_rpc_client(std::string name,
+                                                    std::shared_ptr<grpc::Channel> chan) override;
+    BaseSubtype(const google::protobuf::ServiceDescriptor* service_descriptor)
+        : ResourceSubtype(service_descriptor){};
+};
+
+/// @class Base base.hpp "components/base/base.hpp"
+/// @brief A `Base` is the platform that the other parts of a mobile robot attach to.
+/// @ingroup Base
+///
+/// This acts as an abstract parent class to be inherited from by any drivers representing
+/// specific base implementations. This class cannot be used on its own.
+class Base : public ComponentBase {
+   public:
+    // functions shared across all components
+    static std::shared_ptr<ResourceSubtype> resource_subtype();
+    static Subtype subtype();
+
+    /// @brief Move a robot's base in a straight line by a given distance. This method blocks
+    /// until completed or cancelled
+    /// @param distance_mm Desired travel distance in millimeters
+    /// @param mm_per_sec Desired travel velocity in millimeters/second
+    virtual void move_straight(int64_t distance_mm, double mm_per_sec) = 0;
```

Comment:
[q] Right now we do not use `boost::units`. There was an open question about that in the Servo PR. I have left it like this for now but I am happy to swap if that is the more modern approach.

Reply:
I don't have any immediate thoughts about this. Will do a bit of research/thinking, but in the meantime I defer to @acmorrow